File: ruby/lib/runapi/seedream/resources/decompose_layers.rb
# frozen_string_literal: true

module RunApi
  module Seedream
    module Resources
      # Separates one image into a base image and independent layers.
      class DecomposeLayers
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/seedream/decompose_layers"
        RESPONSE_CLASS = Types::DecomposeLayersResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedDecomposeLayersResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_contract!(CONTRACT["decompose-layers"], params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end
      end
    end
  end
end

File: python/src/runapi/seedream/resources/decompose_layers.py
"""Seedream layer decomposition resource."""

from __future__ import annotations

from typing import Any, Optional

from runapi.core import RequestOptions, Resource

from ..contract_gen import CONTRACT
from ..types import CompletedDecomposeLayersResponse, DecomposeLayersResponse


class DecomposeLayers(Resource):
    """Separate one image into a base image and independent layers."""

    ENDPOINT = "/api/v1/seedream/decompose_layers"
    RESPONSE_CLASS = DecomposeLayersResponse
    COMPLETED_RESPONSE_CLASS = CompletedDecomposeLayersResponse

    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))

    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        compacted = self._compact_params(params)
        self._validate_contract(CONTRACT["decompose-layers"], compacted)
        return self._request("post", self.ENDPOINT, body=compacted, options=options)

    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)
